File: crates/quilltap-core/src/announcement_attribution.rs
//! Port of v4's `lib/chat/context/announcement-attribution.ts` — speaker
//! attribution for ad-hoc announcements (the Insert Announcement composer).
//!
//! `customAnnouncer` names who is speaking — an off-scene workspace character, or
//! a free-text display name — but it is a *rendering* field: the Salon paints that
//! name and avatar on the bubble, and nothing carried it into the model's context.
//! So an announcement posted as a named character arrived at every character as an
//! anonymous block of prose, and the model had to guess who was talking. It
//! guesses badly, and confidently: a whispered announcement written in one
//! character's voice was read as a different character entirely, and the mistake
//! then became part of the scene.
//!
//! Every other authored line in a context carries a speaker — participant messages
//! are tagged `[Name]` by [`crate::message_attribution::attribute_messages_for_character`],
//! and Staff name themselves in their own prose ("Prospero opens his ledger…").
//! The ad-hoc announcement was the sole anonymous line, and it is the one place
//! the operator explicitly *chose* a speaker. The Courier transport already
//! resolved the name for its own transcript
//! ([`crate::services::courier_transport`]), so the same message was attributed on
//! one path and anonymous on the other — the model's reading of a scene changed
//! with the transport carrying it.
//!
//! The tag uses the same `[Name] ` form the multi-character path already emits, so
//! it reads as one convention rather than a second dialect.
//!
//! **Pure** — no repository access, so the name map is the caller's problem
//! (`services::message_context` builds it). Pinned by
//! `announcement_attribution_equivalence` (tier 1, exact).

use std::collections::HashMap;

use serde_json::Value;

use crate::jsstr::js_trim;
use crate::staff_display_names::staff_display_name;

/// The `customAnnouncer` column's shape, structural so callers stay decoupled
/// (v4 `CustomAnnouncer`). `kind` is carried as the raw string: v4 types it
/// `'character' | 'custom'` but tests only `=== 'character'` at runtime, so any
/// other stored value takes the display-name branch.
#[derive(Clone, Debug, PartialEq, Eq, Default)]
pub struct CustomAnnouncer {
    pub kind: String,
    pub character_id: Option<String>,
    pub display_name: Option<String>,
}

impl CustomAnnouncer {
    /// Read the stored `customAnnouncer` JSON. A non-object (absent, null, or a
    /// scalar) is `None` — v4's optional chaining reaches the same place.
    pub fn from_value(v: Option<&Value>) -> Option<CustomAnnouncer> {
        let o = v?.as_object()?;
        Some(CustomAnnouncer {
            kind: o
                .get("kind")
                .and_then(Value::as_str)
                .unwrap_or_default()
                .to_string(),
            character_id: o
                .get("characterId")
                .and_then(Value::as_str)
                .map(str::to_string),
            display_name: o
                .get("displayName")
                .and_then(Value::as_str)
                .map(str::to_string),
        })
    }
}

/// v4's structural `AnnouncerAttributable` — the fields the pass reads, plus the
/// `{...m, content}` / `{...m, opaqueContent}` spreads it uses to rewrite them.
pub trait AnnouncerAttributable: Clone {
    fn content(&self) -> Option<&str>;
    /// Returned by value so a JSON-shaped carrier can read the column on demand
    /// (the announcer bag is three short strings and only announcement rows carry
    /// one, so the copy is not worth a lifetime).
    fn custom_announcer(&self) -> Option<CustomAnnouncer>;
    /// `typeof m.opaqueContent === 'string'` — `Some` only when the column is a
    /// string (an absent or null `opaqueContent` is `None`, matching v4's guard).
    fn opaque_content(&self) -> Option<&str>;
    /// The `systemSender` column (the Staff signer) — `None` when absent/null/"".
    fn system_sender(&self) -> Option<&str>;
    /// The `systemKind` column; the staff fallback fires only for `"announcement"`.
    fn system_kind(&self) -> Option<&str>;
    /// v4 `{ ...m, content }` — every other field survives untouched.
    fn with_content(&self, content: String) -> Self;
    /// v4 `next.opaqueContent = …` — rewrites only `opaqueContent`. Separate from
    /// [`Self::with_content`] because the pass writes both, independently.
    fn with_opaque_content(&self, opaque_content: String) -> Self;
}

/// Resolve the display name for an announcer, or `None` when it can't be named
/// (v4 `resolveAnnouncerName`).
///
/// A `character` announcer whose id resolves to nothing — deleted since the
/// announcement was posted — returns `None` rather than a placeholder: a wrong or
/// invented name is worse than no name, because the model treats a name as fact.
///
/// When no `customAnnouncer` is present the announcement was signed as Staff (the
/// Insert Announcement dialog's `staff` mode writes a `systemSender` and no
/// `customAnnouncer`). Those still need a speaker: an operator-authored line signed
/// as the Host is not prose the Host wrote, so it carries no self-naming, and
/// without a fallback it reaches the model as an anonymous `user` turn. Fall back
/// to the `systemSender`, resolved through the single staff-name table.
pub fn resolve_announcer_name(
    announcer: Option<CustomAnnouncer>,
    character_names_by_id: &HashMap<String, String>,
    system_sender: Option<&str>,
) -> Option<String> {
    if let Some(announcer) = announcer {
        if announcer.kind == "character" {
            // `if (!announcer.characterId) return null` — the empty string is falsy.
            let id = announcer
                .character_id
                .as_deref()
                .filter(|s| !s.is_empty())?;
            return character_names_by_id
                .get(id)
                .map(|n| js_trim(n).to_string())
                .filter(|n| !n.is_empty());
        }

        return announcer
            .display_name
            .as_deref()
            .map(|n| js_trim(n).to_string())
            .filter(|n| !n.is_empty());
    }

    // v4 `if (systemSender) return staffDisplayName(systemSender).trim() || null`.
    // `if (systemSender)` is JS-truthy, so "" is excluded but a whitespace-only
    // sender is NOT — it reaches `staffDisplayName` (returns the raw tag) and the
    // `.trim() || null` then nulls it out.
    let sender = system_sender.filter(|s| !s.is_empty())?;
    let name = js_trim(&staff_display_name(Some(sender))).to_string();
    (!name.is_empty()).then_some(name)
}

/// Character ids an announcement references, for a single up-front name lookup
/// (v4 `collectAnnouncerCharacterIds`). First-appearance order, deduped.
pub fn collect_announcer_character_ids<T: AnnouncerAttributable>(messages: &[T]) -> Vec<String> {
    let mut seen = std::collections::HashSet::new();
    let mut ids = Vec::new();
    for m in messages {
        if let Some(a) = m.custom_announcer() {
            if a.kind == "character" {
                if let Some(id) = a.character_id.as_deref().filter(|s| !s.is_empty()) {
                    if seen.insert(id.to_string()) {
                        ids.push(id.to_string());
                    }
                }
            }
        }
    }
    ids
}

/// Prefix each ad-hoc announcement's body with its speaker (v4
/// `attributeAdhocAnnouncements`).
///
/// A `customAnnouncer` (character/custom mode) names the speaker directly. A
/// `staff`-mode announcement carries a `systemSender` instead — but only ad-hoc
/// announcements (`systemKind === 'announcement'`) take that fallback: ordinary
/// Staff whispers (image notices, tool bubbles, memory recalls) also carry a
/// `systemSender` and name themselves in their own prose, so prefixing them here
/// would double-tag every one. A message with neither field, and an announcer that
/// can't be named, both pass through unchanged.
///
/// The prefix lands on `opaqueContent` too when present: an opaque-anywhere chat
/// swaps that persona-free body into the LLM context in place of `content`
/// ([`crate::services::message_context::normalize_whisper_roles`]), so tagging only
/// `content` would leave the model's copy anonymous in exactly that mode.
pub fn attribute_adhoc_announcements<T: AnnouncerAttributable>(
    messages: &[T],
    character_names_by_id: &HashMap<String, String>,
) -> Vec<T> {
    messages
        .iter()
        .map(|m| {
            // v4 `const systemSender = m.systemKind === 'announcement' ? m.systemSender : undefined`.
            let system_sender = if m.system_kind() == Some("announcement") {
                m.system_sender()
            } else {
                None
            };
            let Some(name) =
                resolve_announcer_name(m.custom_announcer(), character_names_by_id, system_sender)
            else {
                return m.clone();
            };
            let tag = format!("[{name}]");
            // Idempotent: re-running (a retry, a regenerate) must not stack tags.
            let prefix = |text: &str| {
                if text.starts_with(&tag) {
                    text.to_string()
                } else {
                    format!("{tag} {text}")
                }
            };
            // v4 `{ ...m, content: prefix(m.content ?? '') }`.
            let mut next = m.with_content(prefix(m.content().unwrap_or("")));
            // v4 `if (typeof m.opaqueContent === 'string') next.opaqueContent = prefix(m.opaqueContent)`.
            if let Some(opaque) = m.opaque_content() {
                next = next.with_opaque_content(prefix(opaque));
            }
            next
        })
        .collect()
}

/// v4's interface is structural, so a raw message object satisfies it directly.
/// This impl is what the differential drives (and what any JSON-shaped caller
/// gets for free); the production caller uses
/// [`crate::services::message_context::WhisperMessage`].
impl AnnouncerAttributable for Value {
    fn content(&self) -> Option<&str> {
        self.get("content").and_then(Value::as_str)
    }
    fn custom_announcer(&self) -> Option<CustomAnnouncer> {
        CustomAnnouncer::from_value(self.get("customAnnouncer"))
    }
    fn opaque_content(&self) -> Option<&str> {
        // `typeof === 'string'`: a null/absent opaqueContent is not a string.
        self.get("opaqueContent").and_then(Value::as_str)
    }
    fn system_sender(&self) -> Option<&str> {
        self.get("systemSender").and_then(Value::as_str)
    }
    fn system_kind(&self) -> Option<&str> {
        self.get("systemKind").and_then(Value::as_str)
    }
    fn with_content(&self, content: String) -> Value {
        let mut v = self.clone();
        if let Some(o) = v.as_object_mut() {
            o.insert("content".to_string(), Value::String(content));
        }
        v
    }
    fn with_opaque_content(&self, opaque_content: String) -> Value {
        let mut v = self.clone();
        if let Some(o) = v.as_object_mut() {
            o.insert("opaqueContent".to_string(), Value::String(opaque_content));
        }
        v
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn names() -> HashMap<String, String> {
        HashMap::from([("ariel".to_string(), "Ariel".to_string())])
    }

    #[test]
    fn unresolvable_character_is_never_invented() {
        let gone = CustomAnnouncer {
            kind: "character".into(),
            character_id: Some("gone".into()),
            display_name: None,
        };
        assert_eq!(resolve_announcer_name(Some(gone), &names(), None), None);
        assert_eq!(resolve_announcer_name(None, &names(), None), None);
    }

    #[test]
    fn staff_sender_is_the_fallback_speaker() {
        // No announcer, a Staff signer → the staff display name (bug 28).
        assert_eq!(
            resolve_announcer_name(None, &names(), Some("host")),
            Some("The Host".to_string())
        );
        // A whitespace-only sender is JS-truthy but trims empty → None.
        assert_eq!(resolve_announcer_name(None, &names(), Some("   ")), None);
        // An announcer, when present, wins over the sender.
        let ariel = CustomAnnouncer {
            kind: "character".into(),
            character_id: Some("ariel".into()),
            display_name: None,
        };
        assert_eq!(
            resolve_announcer_name(Some(ariel), &names(), Some("host")),
            Some("Ariel".to_string())
        );
    }
}

File: harness/oracle/cases/announcement-attribution.ts
/**
 * Oracle case (P4.D37, v4 `424a7381`): the pure ad-hoc-announcement speaker
 * attribution leaves.
 *
 * Drives the REAL exported functions from v4's
 * `lib/chat/context/announcement-attribution.ts`:
 *   - resolveAnnouncerName
 *   - collectAnnouncerCharacterIds
 *   - attributeAdhocAnnouncements
 *
 * The corpus is v4's own 109-line jest suite (tagging a character / a custom
 * display name, an un-announced message left alone, an unresolvable character
 * passed through unnamed, idempotency across a re-run, every other field
 * preserved) widened where the suite relies on TypeScript rather than a runtime
 * check: a blank/whitespace display name, an empty-string `characterId`, a name
 * that needs trimming, an absent `content`, a body that starts with the tag but a
 * DIFFERENT name, and the first-appearance ordering of the id collection.
 *
 * `attributeAdhocAnnouncements` is emitted as the WHOLE output message array, so
 * "every other field survives" is diffed rather than asserted.
 *
 * Run from inside the server checkout:
 *   cd ~/source/quilltap-server
 *   npx tsx <v5-worktree>/harness/oracle/cases/announcement-attribution.ts \
 *     > /tmp/oracle-announcement-attribution.ndjson
 */

import {
  attributeAdhocAnnouncements,
  collectAnnouncerCharacterIds,
  resolveAnnouncerName,
} from '@/lib/chat/context/announcement-attribution';

type Msg = Record<string, unknown>;

const ARIEL = '14710fac-c074-45f1-bc03-c19fa66fc18e';
const BEA = '2b2b2b2b-0000-4000-8000-000000000002';

/** The shared name map for every case (the caller's up-front lookup). */
const NAME_ENTRIES: Array<[string, string]> = [
  [ARIEL, 'Ariel'],
  [BEA, '  Bea  '], // stored untrimmed: v4 trims at resolve time, not at set time
  ['blank-name', '   '],
  // A name under the EMPTY-STRING key, so v4's `!announcer.characterId` falsy
  // guard is observable: without it, an empty `characterId` would resolve to a
  // name. (Without this entry the guard is a no-op the map already covers, and
  // the arm cannot be mutation-proven.)
  ['', 'Nobody At All'],
];
const NAMES = new Map(NAME_ENTRIES);

const lines: string[] = [];

// ---------------------------------------------------------------------------
// resolveAnnouncerName
// ---------------------------------------------------------------------------

const resolveCases: Array<{ name: string; announcer: unknown; systemSender?: string | null }> = [
  { name: 'character_named', announcer: { kind: 'character', characterId: ARIEL } },
  { name: 'character_name_is_trimmed', announcer: { kind: 'character', characterId: BEA } },
  { name: 'character_name_blank_is_null', announcer: { kind: 'character', characterId: 'blank-name' } },
  { name: 'character_unknown_id_is_null', announcer: { kind: 'character', characterId: 'gone' } },
  { name: 'character_null_id_is_null', announcer: { kind: 'character', characterId: null } },
  { name: 'character_empty_id_is_null', announcer: { kind: 'character', characterId: '' } },
  { name: 'character_absent_id_is_null', announcer: { kind: 'character' } },
  { name: 'custom_display_name', announcer: { kind: 'custom', displayName: 'The Narrator' } },
  { name: 'custom_display_name_trimmed', announcer: { kind: 'custom', displayName: '  The Narrator  ' } },
  { name: 'custom_blank_display_name_is_null', announcer: { kind: 'custom', displayName: '   ' } },
  { name: 'custom_absent_display_name_is_null', announcer: { kind: 'custom' } },
  { name: 'announcer_null', announcer: null },
  { name: 'announcer_undefined', announcer: undefined },
  // Bug 28 (v4 `99d5fc7d`): the Staff-sender fallback. No `customAnnouncer`, a
  // `systemSender` names the speaker through the staff-name table.
  { name: 'staff_sender_host', announcer: null, systemSender: 'host' },
  { name: 'staff_sender_suparna', announcer: null, systemSender: 'suparna' },
  // An unknown sender falls back to the raw tag (staffDisplayName `?? sender`).
  { name: 'staff_sender_unknown_is_raw', announcer: null, systemSender: 'newcomer' },
  // JS-truthy but whitespace-only → staffDisplayName returns the raw tag, `.trim()`
  // empties it, `|| null` nulls it.
  { name: 'staff_sender_whitespace_is_null', announcer: null, systemSender: '   ' },
  // Empty string is falsy → the `if (systemSender)` guard is skipped → null.
  { name: 'staff_sender_empty_is_null', announcer: null, systemSender: '' },
  // An announcer, when present, wins over the sender (returns from the announcer
  // branch before the fallback is reached).
  { name: 'announcer_wins_over_sender', announcer: { kind: 'character', characterId: ARIEL }, systemSender: 'host' },
];

for (const c of resolveCases) {
  lines.push(
    JSON.stringify({
      kind: 'resolve',
      name: c.name,
      announcer: c.announcer ?? null,
      systemSender: c.systemSender ?? null,
      names: NAME_ENTRIES,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      output: resolveAnnouncerName(c.announcer as never, NAMES, c.systemSender ?? undefined),
    }),
  );
}

// ---------------------------------------------------------------------------
// collectAnnouncerCharacterIds
// ---------------------------------------------------------------------------

const collectCases: Array<{ name: string; messages: Msg[] }> = [
  {
    name: 'gathers_referenced_ids_once_each',
    messages: [
      { customAnnouncer: { kind: 'character', characterId: ARIEL } },
      { customAnnouncer: { kind: 'character', characterId: ARIEL } },
      { customAnnouncer: { kind: 'custom', displayName: 'The Narrator' } },
      { customAnnouncer: null },
      {},
    ],
  },
  {
    name: 'first_appearance_order',
    messages: [
      { customAnnouncer: { kind: 'character', characterId: BEA } },
      { customAnnouncer: { kind: 'character', characterId: ARIEL } },
      { customAnnouncer: { kind: 'character', characterId: BEA } },
    ],
  },
  {
    name: 'skips_empty_and_null_ids',
    messages: [
      { customAnnouncer: { kind: 'character', characterId: '' } },
      { customAnnouncer: { kind: 'character', characterId: null } },
      { customAnnouncer: { kind: 'character' } },
    ],
  },
  { name: 'empty_list', messages: [] },
];

for (const c of collectCases) {
  lines.push(
    JSON.stringify({
      kind: 'collect',
      name: c.name,
      messages: c.messages,
      output: collectAnnouncerCharacterIds(c.messages as never),
    }),
  );
}

// ---------------------------------------------------------------------------
// attributeAdhocAnnouncements
// ---------------------------------------------------------------------------

const attributeCases: Array<{ name: string; messages: Msg[]; twice?: boolean }> = [
  {
    name: 'tags_a_character_announcement',
    messages: [
      { content: '*She drifts closer.*', customAnnouncer: { kind: 'character', characterId: ARIEL } },
    ],
  },
  {
    name: 'tags_a_custom_announcement',
    messages: [{ content: 'A bell tolls.', customAnnouncer: { kind: 'custom', displayName: 'The Narrator' } }],
  },
  {
    name: 'leaves_unannounced_messages_untouched',
    messages: [{ content: 'Prospero opens his ledger.' }, { content: 'hi', customAnnouncer: null }],
  },
  {
    name: 'leaves_an_unresolvable_character_alone',
    messages: [{ content: 'x', customAnnouncer: { kind: 'character', characterId: 'gone' } }],
  },
  {
    name: 'idempotent_across_a_rerun',
    twice: true,
    messages: [{ content: 'hello', customAnnouncer: { kind: 'character', characterId: ARIEL } }],
  },
  {
    name: 'a_different_name_in_brackets_is_not_the_tag',
    messages: [
      { content: '[Bea] hello', customAnnouncer: { kind: 'character', characterId: ARIEL } },
    ],
  },
  {
    name: 'preserves_every_other_field',
    messages: [
      {
        content: 'hi',
        customAnnouncer: { kind: 'character', characterId: ARIEL },
        id: 'm1',
        role: 'ASSISTANT',
        targetParticipantIds: ['p1'],
        systemKind: 'announcement',
        participantId: null,
      },
    ],
  },
  {
    name: 'absent_content_becomes_the_bare_tag',
    messages: [{ id: 'm2', customAnnouncer: { kind: 'custom', displayName: 'The Narrator' } }],
  },
  {
    name: 'a_mixed_list_in_one_pass',
    messages: [
      { content: 'u01', role: 'USER' },
      { content: 'a01', role: 'ASSISTANT', customAnnouncer: { kind: 'character', characterId: ARIEL } },
      { content: 'a02', role: 'ASSISTANT', customAnnouncer: { kind: 'custom', displayName: 'A Distant Bell' } },
      { content: 'a03', role: 'ASSISTANT', customAnnouncer: { kind: 'character', characterId: 'gone' } },
    ],
  },
  // ---- Bug 28: the Staff-sender fallback + the opaque body ----
  {
    // A staff-signed ad-hoc announcement (no customAnnouncer, systemKind
    // 'announcement') is tagged with the resolved staff name.
    name: 'tags_a_staff_announcement',
    messages: [
      { content: 'The hour grows late.', systemSender: 'host', systemKind: 'announcement' },
    ],
  },
  {
    // An ordinary Staff whisper (a NON-announcement systemKind) carries a
    // systemSender too, but names itself in its own prose — it must NOT be tagged.
    name: 'leaves_non_announcement_staff_whispers_untouched',
    messages: [
      { content: 'A Lantern image was shared.', systemSender: 'lantern', systemKind: 'image' },
    ],
  },
  {
    // The prefix lands on opaqueContent as well as content, so the opaque-anywhere
    // body swap (normalizeWhisperRoles) still reaches the model attributed.
    name: 'tags_both_content_and_opaque_content',
    messages: [
      {
        content: 'A bell tolls thrice.',
        opaqueContent: 'A bell tolls thrice.',
        systemSender: 'host',
        systemKind: 'announcement',
      },
    ],
  },
  {
    // A null opaqueContent is not a string — only content is tagged.
    name: 'null_opaque_content_is_left_alone',
    messages: [
      {
        content: 'Only content here.',
        opaqueContent: null,
        customAnnouncer: { kind: 'custom', displayName: 'The Narrator' },
      },
    ],
  },
  {
    // Idempotent across a re-run for the staff arm too (retry / regenerate).
    name: 'idempotent_staff_across_a_rerun',
    twice: true,
    messages: [
      {
        content: 'The doors close.',
        opaqueContent: 'The doors close.',
        systemSender: 'suparna',
        systemKind: 'announcement',
      },
    ],
  },
];

for (const c of attributeCases) {
  const once = attributeAdhocAnnouncements(c.messages as never, NAMES);
  const out = c.twice ? attributeAdhocAnnouncements(once, NAMES) : once;
  lines.push(
    JSON.stringify({
      kind: 'attribute',
      name: c.name,
      twice: c.twice ?? false,
      names: NAME_ENTRIES,
      messages: c.messages,
      output: out,
    }),
  );
}

for (const l of lines) process.stdout.write(l + '\n');
